File: app/src/main/java/com/twitter/challenge/view/ui/MainActivity.java
package com.twitter.challenge.view.ui;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProviders;

import com.twitter.challenge.R;
import com.twitter.challenge.databinding.ActivityMainBinding;
import com.twitter.challenge.viewmodel.WeatherViewModel;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private final String TAG = this.getClass().getSimpleName();

    private ActivityMainBinding binding;
    private WeatherViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        viewModel = ViewModelProviders.of(this).get(WeatherViewModel.class);
        viewModel.loadCurrentWeatherConditions();

        binding.setViewModel(viewModel);
        binding.setMainActivity(this);
        binding.setLifecycleOwner(this);
    }

    public void onClick(View v) {
        Log.d(TAG, "onClick: v = " + v.getId());
        if (v.getId() == R.id.deviation_btn) {
            viewModel.loadFutureWeatherConditions();
        }
    }
}

File: app/src/main/java/com/twitter/challenge/view/ui/MainActivity.kt
package com.twitter.challenge.view.ui

import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.twitter.challenge.R
import com.twitter.challenge.viewmodel.WeatherViewModel
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {

    private val TAG = this.javaClass.simpleName

    @Inject
    lateinit var factory: MainFragmentFactory

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        supportFragmentManager.fragmentFactory = factory
        addLoadingFragment()

        val viewModel: WeatherViewModel by viewModels()
        viewModel.loadCurrentWeatherConditions()

        viewModel.loadingComplete.observe(this, Observer {
            Log.d(TAG, "onCreate: loading complete observed : $it")
            addWeatherFragment()
        })
    }

    private fun addLoadingFragment() {
        supportFragmentManager.beginTransaction()
                .replace(R.id.main_fragment_container, LoadingFragment::class.java, null)
                .commit()
    }

    private fun addWeatherFragment() {
        supportFragmentManager.beginTransaction()
                .replace(R.id.main_fragment_container, WeatherFragment::class.java, null)
                .commit()
    }
}